### PR TITLE
dev-cmd/bottle: make gzip creation consistent across platforms

### PR DESCRIPTION
**Pull request by Bo98**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Apple gzip and GNU gzip differ in implementation and produce different gzips.

Instead, we can use the Ruby's `Zlib::GzipWriter` to produce consistent output across all platforms.

Note that the gzips produced here actually match what Apple gzip does, as Apple gzip also uses zlib internally. So this change should not result in any checksum change to macOS bottles.

### CODE REVIEW

File: Library/Homebrew/dev-cmd/bottle.rb [hunk 1]
```diff
@@ -441,9 +442,12 @@ def bottle_formula(f, args:)
           mv tar_path, relocatable_tar_path
           # Use gzip, faster to compress than bzip2, faster to uncompress than bzip2
           # or an uncompressed tarball (and more bandwidth friendly).
-          safe_system "gzip", "-f", relocatable_tar_path
+          gz = Zlib::GzipWriter.open(bottle_path)
+          gz.mtime = tab.source_modified_time
+          gz.orig_name = relocatable_tar_path
+          gz.write File.binread(relocatable_tar_path)
```
Comment (Bo98) [inline Library/Homebrew/dev-cmd/bottle.rb:448]: This is perhaps not RAM efficient - it reads the entire `tar` into memory and then writes it to gzip in one go.

I'll investigate chunking this read into smaller segments.

**Comment by MikeMcQuaid** [comment]
> Note that the gzips produced here actually match what Apple gzip does, as Apple gzip also uses zlib internally. So this change should not result in any checksum change to macOS bottles.

Given this: could/should this logic be scoped just to Linux? Is there something we could `brew install` instead? Would be interested in a benchmark of one of our larger bottles and how the performance differs here.

**Comment by Bo98** [comment]
> could/should this logic be scoped just to Linux?

I do not see any benefit achieved from doing this.

It's probably actually a little safer to use the same code in case there is some differences in implementations based on buffer sizes. The sample size of my tests comparing the output to Apple gzip is 1.

> Is there something we could `brew install` instead?

I am not aware of a formula with a gzip implementation that uses zlib. GNU gzip could be installed on macOS, though that would cause a change for all existing bottle checksums.

I also don't see a reason to avoid using the zlib we build into portable Ruby. We might as well strip it out if we don't want to ever use it.

> Would be interested in a benchmark of one of our larger bottles and how the performance differs here.

This is the interesting question and I will report back on this.

**Comment by MikeMcQuaid** [comment]
> I do not see any benefit achieved from doing this.

Some benefits:
- we avoid changing the bottling behaviour on macOS that has served us well for a long time
- we shell out to an external tool rather than configuring it ourselves

I don't think these are strong arguments in favour of the status quo but are worth mentioning anyway.

> It's probably actually a little safer to use the same code in case there is some differences in implementations based on buffer sizes or something like that. The sample size of my tests comparing the output to Apple gzip is 1.

Ok. It'd be good to verify this with a lot more cases.

> I am not aware of a formula with a gzip implementation that uses zlib. GNU gzip could be installed on macOS, though that would cause a change for all existing bottle checksums.

I'd like us to have a lot more confidence that this code will not change any macOS checksums before discounting the GNU gzip option.

> I also don't see a reason to avoid using the zlib we build into portable Ruby. We might as well strip it out if we don't want to ever use it.

Somewhat off topic but: we want portable Ruby to be as close to system macOS Ruby as possible, regardless of whether we use the functionality or not.

> This is the interesting question and I will report back on this.

Thanks. If this is dramatically faster/slower I think that adds weight to arguments for/against this approach.

**Comment by MikeMcQuaid** [comment]
> Tested on a Swift bottle. This is an extreme case of 3GB. I don't know any other bottle much bigger than 1GB.
> 
> Apple gzip takes ~112s.
> Ruby gzip with 64KB buffer (which I believe matches Apple gzip) takes ~116s. This does get faster with a larger buffer, at the cost of more RAM usage.

Awesome. :shipit:, that's definitely fine.

**Review by MikeMcQuaid — APPROVED** [review]

**Review by MikeMcQuaid — APPROVED** [review]

**Pull request merged**